### PR TITLE
lapack/{gonum,testlapack}: clean up Dgetc2 and Dgesc2 and rework their tests

### PR DESCRIPTION
**Pull request by vladimir-ch**
Please take a look.

cc @soypat

### CODE REVIEW

**Review by kortschak — APPROVED** [review]

File: lapack/gonum/dgetc2.go [hunk 1]
```diff
@@ -52,22 +56,23 @@ func (impl Implementation) Dgetc2(n int, a []float64, lda int, ipiv, jpiv []int)
 		eps    = dlamchP
 		smlnum = dlamchS / eps
 	)
+
 	if n == 1 {
 		ipiv[0], jpiv[0] = 0, 0
 		if math.Abs(a[0]) < smlnum {
-			a[0] = smlnum
 			k = 0
+			a[0] = smlnum
```
Comment (soypat) [inline lapack/gonum/dgetc2.go:64]: Why the switcheroo?

Comment (kortschak) [reply lapack/gonum/dgetc2.go:64]: It mirrors the reference Fortran.

File: lapack/testlapack/dgetc2.go [hunk 2]
```diff
@@ -13,86 +13,123 @@ import (
...
-	// Apply Permutations P and Q to L*U.
+	// Apply permutation matrices P and Q to L*U.
 	for i := n - 1; i >= 0; i-- {
-		ipv, jpv := ipiv[i], jpiv[i]
+		ipv := ipiv[i]
 		if ipv != i {
-			bi.Dswap(n, work.Data[i*lda:], 1, work.Data[ipv*lda:], 1)
+			row1 := blas64.Vector{N: n, Data: diff.Data[i*diff.Stride:], Inc: 1}
+			row2 := blas64.Vector{N: n, Data: diff.Data[ipv*diff.Stride:], Inc: 1}
+			blas64.Swap(row1, row2)
```
Comment (soypat) [inline lapack/testlapack/dgetc2.go:115]: I'm not a fan of using Swap over Dswap.

**Comment by soypat** [comment]
LGTM. All around great improvements save a couple things I might not be used to yet.

File: lapack/testlapack/dgetc2.go (hunk 2 above)
Comment (vladimir-ch) [reply lapack/testlapack/dgetc2.go:115]: Yeah, it's a bit verbose but on the other hand it forces me to have `row1`, `col1`, ... variables which makes the code more explicit and self-documenting. And all the BLAS calls are made through `blas64` package level functions.

File: lapack/testlapack/dgesc2.go [hunk 3]
```diff
@@ -12,80 +12,92 @@ import (
...
+
+	// Check that const input to Dgesc2 hasn't been modified.
+	if !floats.Same(lu.Data, luCopy.Data) {
+		t.Errorf("%v: unexpected modification in lu", name)
+	}
+	if !intsEqual(ipiv, ipivCopy) {
+		t.Errorf("%v: unexpected modification in ipiv", name)
+	}
+	if !intsEqual(jpiv, jpivCopy) {
+		t.Errorf("%v: unexpected modification in ipiv", name)
```
Comment (soypat) [inline lapack/testlapack/dgesc2.go:80]: "jpiv"

Comment (vladimir-ch) [reply lapack/testlapack/dgesc2.go:80]: Thanks, fix submitted in https://github.com/gonum/gonum/pull/1675.

**Pull request merged**